File: lymui/grayscale.c
//
//  grayscale.c
//  lymui
//
//  Created by Marc Intha on 06/01/2019.
//  Copyright © 2019 Marc. All rights reserved.
//

#include <stdlib.h>
#include "grayscale.h"
#include "helper.h"

Gray getGrayScale(Rgb *rgb, Strategy s) {
    if (rgb == NULL) {
        return 0;
    }
    
    double _r = (double) rgb->r;
    double _g = (double) rgb->g;
    double _b = (double) rgb->b;
    double grayscale = 0.0;
    
    switch (s) {
        case Lightness:
            grayscale = (getMinMax(_r, _g, _b, 1) + getMinMax(_r, _g, _b, 0)) / 2.0;
            break;
        case Average:
            grayscale = (_r + _g + _b) / 3.0;
            break;
        case bt709:
            grayscale = 0.2126 * _r + 0.7152 * _g + 0.0722 * _b;
            break;
        case bt2100:
            grayscale = 0.2627 * _r + 0.6780 * _g + 0.0593 * _b;
            break;
        default:
            grayscale = 0.21 * _r + 0.72 * _g + 0.07 * _b;
            break;
    }
    
    return doubleToUint(grayscale);
}

File: lymui/src/grayscale.rs
use crate::rgb::{FromRgb, Rgb};
use crate::util::AsFloat;

#[cfg(feature = "js")]
use crate::js::prelude::*;

#[cfg_attr(feature = "js", derive(FromJsObj, IntoJsObject))]
#[derive(Debug)]
pub struct GrayScale(pub u8);

pub enum Kind {
    Lightness,
    Average,
    Luminosity,
    BT709,
    BT2100,
}

impl FromRgb<Kind> for GrayScale {
    fn from_rgb(rgb: Rgb, kind: Kind) -> Self {
        let (r, g, b) = rgb.as_f64();

        let grayscale = match kind {
            Kind::Lightness => {
                let (min, max) = rgb.get_min_max();
                (min + max) / 2.0
            }
            Kind::Average => (r + g + b) / 3.0,
            Kind::Luminosity => 0.21 * r + 0.72 * g + 0.07 * b,
            Kind::BT709 => 0.2126 * r + 0.7152 * g + 0.0722 * b,
            Kind::BT2100 => 0.2627 * r + 0.6780 * g + 0.0593 * b,
        };

        GrayScale(grayscale as u8)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expect_to_create_grayscale_lightness() {
        let rgb = Rgb { r: 255, g: 0, b: 0 };

        let g = GrayScale::from_rgb(rgb, Kind::Lightness);
        assert_eq!(g.0, 127);
    }

    #[test]
    fn expect_to_create_grayscale_average() {
        let rgb = Rgb { r: 255, g: 0, b: 0 };

        let g = GrayScale::from_rgb(rgb, Kind::Average);
        assert_eq!(g.0, 85);
    }

    #[test]
    fn expect_to_create_grayscale_luminosity() {
        let rgb = Rgb { r: 255, g: 0, b: 0 };

        let g = GrayScale::from_rgb(rgb, Kind::Luminosity);
        assert_eq!(g.0, 53);
    }

    #[test]
    fn expect_to_create_grayscale_bt709() {
        let rgb = Rgb { r: 255, g: 0, b: 0 };

        let g = GrayScale::from_rgb(rgb, Kind::BT709);
        assert_eq!(g.0, 54);
    }

    #[test]
    fn expect_to_create_grayscale_bt2100() {
        let rgb = Rgb { r: 255, g: 0, b: 0 };

        let g = GrayScale::from_rgb(rgb, Kind::BT2100);
        assert_eq!(g.0, 66);
    }
}
